### PR TITLE
PHC Binding - initial contribution

### PR DESCRIPTION
**Pull request by gnlpfjh**
Initial implementation of the PHC modules.
This binding allows you to integrate modules(at the Moment AM, EM and JRM) of PHC, without the PHC control (STM), in openHAB.

### CODE REVIEW

File: addons/binding/org.openhab.binding.phc/ESH-INF/thing/channel-types.xml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        <state readOnly="true"></state>
+    </channel-type>
+
+    <channel-type id="jrm-channel">
+        <item-type>Rollershutter</item-type>
+        <label>PHC JRM Channel</label>
+        <description>Channel to an JRM module.</description>
+    </channel-type>
+ <!-- extra?? -->
+    <channel-type id="window-channel">
```
Comment (kubawolanin) [inline addons/binding/org.openhab.binding.phc/ESH-INF/thing/channel-types.xml]: If it's extra, add `advanced="true"` parameter

File: addons/binding/org.openhab.binding.phc/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -0,0 +1,28 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: PHC Binding
+Bundle-SymbolicName: org.openhab.binding.phc;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.0.0.qualifier
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
+Bundle-ClassPath: .
+Import-Package: 
+ com.google.common.cache;version="10.0.1",
```
Comment (kubawolanin) [inline addons/binding/org.openhab.binding.phc/META-INF/MANIFEST.MF]: Please remove version constraints from the imported packages

File: addons/binding/org.openhab.binding.phc/pom.xml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.openhab.binding</groupId>
+    <artifactId>pom</artifactId>
+    <version>2.0.0-SNAPSHOT</version>
```
Comment (kubawolanin) [inline addons/binding/org.openhab.binding.phc/pom.xml]: Please change to `2.1.0-SNAPSHOT`

File: addons/binding/org.openhab.binding.phc/pom.xml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  </parent>
+
+  <groupId>org.openhab.binding</groupId>
+  <artifactId>org.openhab.binding.phc</artifactId>
+  <version>2.0.0-SNAPSHOT</version>
```
Comment (kubawolanin) [inline addons/binding/org.openhab.binding.phc/pom.xml]: You can remove `version` and `groupId` as it'll be inherited from `parent` node

**Comment by kubawolanin** [comment]
Please also add [a proper about.html file](http://docs.openhab.org/developers/development/guidelines.html#b-osgi-bundles)

**Comment by gnlpfjh** [comment]
I´ve corrected/updated these and some other things.
@kubawolanin

File: addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java [hunk 5]
```diff
@@ -0,0 +1,710 @@
...
+    private static final int SEND_RETRY_TIME = 80; // time to wait for an acknowledge before send the message again
+
+    private InputStream serialIn = null;
+    private OutputStream serialOut = null;
+    short lastReceivedCrc;
+    byte[] messageFragment = null;
+    RXTXPort serialPort = null;
+
+    private Map<String, Boolean> toggleMap = new HashMap<String, Boolean>();
+    private ConcurrentSkipListMap<Long, QueueObject> queue = new ConcurrentSkipListMap<Long, QueueObject>();
```
Comment (gnlpfjh) [inline addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java]: The method subMap (line 278) needs the `ConcurrentSkipListMap`. Maybe you have a better solution.

File: addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java [hunk 6]
```diff
@@ -0,0 +1,710 @@
...
+                serialPort.notifyOnDataAvailable(true);
+                serialPort.setRTS(true);
+
+                // get the output stream
+                serialOut = serialPort.getOutputStream();
+
+                sendPorBroadcast((byte) 0xFF);
+
+                byte[] b = { 0x01 };
+                for (int j = 0; j <= 0x1F; j++) { // TODO: move to PHCHandler?
```
Comment (gnlpfjh) [inline addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java]: I removed the TODO, it isn't sensible.

File: addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java [hunk 7]
```diff
@@ -0,0 +1,710 @@
...
+                            case PHCBindingConstants.CHANNELS_JRM:
+                                sendJRM(qo.getmoduleAddress(), qo.getChannel(), qo.getCommand());
+                                break;
+                        }
+
+                        queue.put(System.currentTimeMillis() + SEND_RETRY_TIME, qo);
+                    } else if (qo.getCounter() >= SEND_RETRY_COUNT
+                            && !qo.getModuleType().equals(PHCBindingConstants.CHANNELS_JRM)) {
+                        // Can´t process the acknowledgement of JRM yet.
+                        // TODO: Thing state to offline?
```
Comment (gnlpfjh) [inline addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java]: I think the serial communication isn't enough reliable. I wrote more detail in the TODO.

File: addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java [hunk 8]
```diff
@@ -0,0 +1,710 @@
...
+            CommPortIdentifier portId = null;
+
+            // parse ports and if the default port is found, initialized the
+            // reader
+            @SuppressWarnings("rawtypes")
+            Enumeration portList = CommPortIdentifier.getPortIdentifiers();
+            while (portList.hasMoreElements()) {
+                CommPortIdentifier id = (CommPortIdentifier) portList.nextElement();
+                if (id.getPortType() == CommPortIdentifier.PORT_SERIAL) {
+                    if (id.getName().equals(port)) {
```
Comment (gnlpfjh) [inline addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java]: I get the name from the configuration. I saw it in another binding (RME Binding).

File: addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java (hunk 5 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java]: The most broad interface you could use then is `ConcurrentNavigableMap` [source](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentSkipListMap.html#subMap(K,%20boolean,%20K,%20boolean))

File: addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java (hunk 8 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java]: I think that example is not the most optimal, our codebase also contains the following which should be prefered (if it works ;-))

```java
CommPortIdentifier.getPortIdentifier(portName);
```

**Comment by martinvw** [comment]
@gnlpfjh thanks for the update, If you are ready with it please let me know.

**Comment by gnlpfjh** [comment]
Thank you, I've added both and it's working well. In addition I've added the configuration and sending for shutter duration.
@martinvw Now i'm ready.

**Comment by martinvw** [comment]
Can you squash the last **7** (!!!!) commits and change commit message to include a `Signed-off-by` (to certify that you wrote it or otherwise have the right to pass it on as an open-source), see http://docs.openhab.org/developers/contributing/contributing.html#sign-your-work

For squashing / rebasing see:
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

**Comment by gnlpfjh** [comment]
@martinvw That doesn't look good. I used reword because the commit update fork is an merge commit too. I have the version bevore rebase on my PC so I can sqashing it(whith the "update fork" commit?).

**Comment by martinvw** [comment]
It does not look good at all, merging is not really compatible with squash/merge and rebase strategy we use in this project, recovering from this is really though. The best thing to do, might be to a make patch for all your changes and apply that to a fresh master. For now it might be better to push previous state as it was.

**Comment by gnlpfjh** [comment]
I would push the state before the rebase. Should I open a new PR and push all changes in one commit? I think that would be the easyest.

**Comment by martinvw** [comment]
Yes and please refer to this one in case we forget :-)

**Comment by gnlpfjh** [comment]
The new clean PR is #2255.

**Pull request closed without merging**